pkg/controller/seed-controller-manager/addoninstaller/addoninstaller_controller.go

```diff
@@ -319,5 +320,17 @@ func skipAddonInstallation(addon kubermaticv1.Addon, cluster *kubermaticv1.Clust
 	if addon.Name == CSIAddonName && cluster.Spec.DisableCSIDriver {
 		return true // skip csi driver installation if DisableCSIDriver is true
 	}
+
+	// skip default storageclass addon if csi addon is disabled.
+	if addon.Name == defaultStorageClassAddonName && cluster.Spec.DisableCSIDriver {
+		csiAddonCondition, ok := cluster.Status.Conditions[kubermaticv1.ClusterConditionCSIAddonInUse]
+		if ok {
+			if csiAddonCondition.Status == corev1.ConditionFalse {
+				return true
+			}
+		}
+		return true
+	}
+
```

Comment:
For everything else, we use `{{ if ... }}` statements in our addons. I think we should keep the same approach and not turn it on its head for this one aspect.

Reply:
I tried the `{{ if ... }}` approach but seeing a couple of issue with that when we disable csi driver on an existing cluster, first is that it doesn't remove the `addon` object in cluster namespace on seed and the controller keeps on reconciling, this also doesn't remove the storage class from the cluster but if we delete it then it doesn't get re-created as the addon manifest becomes empty, this [commit](https://github.com/akash-gautam/kubermatic/commit/6158ab19f147a8c790299d89d726bdddd9c913f5) has the changes for this approach.

Reply:
> first is that it doesn't remove the addon object in cluster namespace on seed and the controller keeps on reconciling

That's not actually a bug, that is the intended behaviour. It's a bit weird, I grant you that :grin: If you have a kubevirt usercluster, you still have the CSI addon, even though the addon contains nothing relevant. Likewise you'd have the CSI addon even if external CCM was disabled. Conditional addon deployment should be done for all of for no addons, doing it for one special case is not a good approach IMHO. And to make that work, you'd need a way for addons to specify their conditions somehow, since the addon controller can also deploy admin-provided addons that we do not know about.

> this also doesn't remove the storage class from the cluster

That is IMHO the actual bug. Resources that disappear from the rendered addon manifest should be pruned, we rely on this implicit cleanup everytime we bump an addon and manifests change because of upstream changes.

Reply:
`Deprecated: kubectl apply will no longer prune non-namespaced resources by default when used with the --namespace flag in a future release. To preserve the current behaviour, list the resources you want to target explicitly in the --prune-allowlist flag.`
Maybe this is the issue, I will try to add the storage class to the "--prune-allowlist" flag and see if that helps.

Reply:
Doesn't help, the apply command with prune is not executed when the manifest is empty, which is the case for "default-storage-class` addon once `DisableCSIDriver` is true.

Reply:
Included the cleanup of storageclass in the addon controller once the manifest for it becomes empty.